### PR TITLE
Jails graphs

### LINKED ISSUE #41: Support graphs in new Jails view
**Is your feature request related to a problem? Please describe.**
As described in the new "Category 2: Jails" section of [the PRD](https://docs.google.com/document/d/1eOiMzk8_eDjC4W0xn89qlkmj_6SOqi1dBTyEusAGGaI/edit?ts=5fb45edc#heading=h.s7ludzwvb7fe), we are going to create a new Jails view alongside the existing Corrections view. The middle section of the view, just below the key insight cards created in #40, is a pair of graphs describing jail incarceration rates.

**Describe the solution you'd like**
The design of these graphs is identical to the graphs from the existing corrections view. But the actual metrics shown are different.

Furthermore, one of these graphs has a new link that opens a modal and lets you select from a list of counties in the state. The functionality for that new button and modal will be handled in a separate ticket (#42), so this ticket should just get the graph operational with "the default" county selected (should be the first one in alphabetical order; see comments in the graphs sub-section of this part of the PRD) and skip the county selector for now.

**Describe alternatives you've considered**
N/A

**Additional context**
See the new "Category 2: Jails" section of [the PRD](https://docs.google.com/document/d/1eOiMzk8_eDjC4W0xn89qlkmj_6SOqi1dBTyEusAGGaI/edit?ts=5fb45edc#heading=h.s7ludzwvb7fe) and the Figma mocks it links to.

### PR DESCRIPTION
**Pull request by nasaownsky**
## Support graphs in new Jails view 

> Added graph support for new Jail view 

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #41 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -96,13 +101,32 @@ const App = ({ stateCode, correctionsData, jailsData }) => {
...
+    ["Statewide", "US_CO_ARAPAHOE"],
+    ["Statewide", "Arapahoe county"],
+    coveredCounty
+  );
+
+  const incarcerationRateTopCountiesChartData = generateJailsChartData(
+    jailsMetricData,
+    INCARCERATION_RATE_JAIL,
+    ["US_CO_DENVER", "US_CO_EL_PASO", "US_CO_ARAPAHOE", "US_CO_JEFFERSON", "US_CO_ADAMS"],
+    ["Denver county", "El Paso county", "Arapahoe county", "Jefferson county", "Adams county"]
```
Comment (jessex) [inline src/App.js:118]: Are you planning to make this dynamic in the current PR or in a follow-up PR, @nasaownsky?

File: src/App.js [hunk 2]
```diff
@@ -96,13 +101,32 @@ const App = ({ stateCode, correctionsData, jailsData }) => {
     releasesChartData.sourceData,
   ]);
 
+  const { jailsKeyInsightsData, coveredCounty } = generateJailsKeyInsightsData(jailsMetricData);
+
+  const incarcerationRateChartData = generateJailsChartData(
+    jailsMetricData,
+    INCARCERATION_RATE_JAIL,
+    ["Statewide", "US_CO_ARAPAHOE"],
+    ["Statewide", "Arapahoe county"],
```
Comment (jessex) [inline src/App.js:110]: You're planning to keep Arapahoe County hard-coded until #42, I'm guessing?

File: src/__tests__/App.test.js [hunk 3]
```diff
@@ -35,8 +35,7 @@ describe("App.js", () => {
   beforeEach(() => {
```
Comment (jessex) [inline src/__tests__/App.test.js:35]: We should update this test class to also assert on jails-related calls below.

File: src/utils/getNormalizedStateData.js [hunk 4]
```diff
@@ -72,7 +72,7 @@ const getNormalizedStateData = (data, stateCode) => {
           month: toInt(monthReported - 1),
           day: toInt(dayReported),
         },
-        value: parseFloat(item.value),
+        value: toInt(item.value),
```
Comment (jessex) [inline src/utils/getNormalizedStateData.js:75]: Why did we make this switch?

File: src/utils/toInt.js [hunk 5]
```diff
@@ -14,6 +14,6 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 // =============================================================================
-const toInt = (string) => parseInt(string, 10);
+const toInt = (string) => parseFloat(string, 10);
```
Comment (jessex) [inline src/utils/toInt.js:17]: And why this opposing switch?

File: src/utils/generateJailsChartData.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
+// Recidiviz - a data platform for criminal justice reform
+// Copyright (C) 2020 Recidiviz, Inc.
```
Comment (jessex) [inline src/utils/generateJailsChartData.js:2]: Whoops: new license text should be 2021 now.

File: src/utils/generateJailsChartData.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+export const noMetricData = (metric) =>
+  `${metric} doesn't appear in source files. Chart data is not generated.`;
+
+/**
+ * Transforms normalized data to chart.js format.
+ * @param data - normalized data (see @returns of getNormalizedStateData)
+ * @param metric - metric name for which we generate chart data
+ * @param counties {string[]} - counties for which we generate chart data
+ * @param countyLabels {string[]} - humanized county names
+ * @param coveredCounty - counties reporting percent
```
Comment (jessex) [inline src/utils/generateJailsChartData.js:30]: Let's change this to be a bit more explicit: `coveredCounty - percentage of counties in the state reporting data`

File: src/utils/generateJailsChartData.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+ *   isNotAvailable: boolean
+ *   isStatewide: boolean
+ *   label: string
+ *   data: (number|null)[]
+ * }[],
+ * labels: string[],
+ * }}
+ */
+
+const generateJailsChartData = (data, metric, counties, countyLabels = [], coveredCounty) => {
```
Comment (jessex) [inline src/utils/generateJailsChartData.js:45]: Let's make sure to add unit tests for this one before we merge.

Comment (jessex) [reply src/utils/generateJailsChartData.js:45]: Also, the logic in this function looks _very similar_ to the logic in `generateCorrectionsChartData`. Let's try to abstract the common parts out into a helper function? At the very least, it looks like the `periods` creation is identical, unless I'm missing something.

File: src/utils/generateJailsKeyInsightsData.js [hunk 9]
```diff
@@ -107,32 +107,38 @@ const generateJailsKeyInsightsData = (data) => {
...
-      return keyInsights;
-    },
-    []
-  );
+    return keyInsights;
+  }, []);
+
+  return {
+    jailsKeyInsightsData,
+    coveredCounty: flowData[PERCENTAGE_COVERED_COUNTY].number,
```
Comment (jessex) [inline src/utils/generateJailsKeyInsightsData.js:140]: I actually think we should rename this var from `coveredCounty` to `countyCoverage` throughout the PR.

**Review by jessex — CHANGES_REQUESTED** [review]
Great work so far, @nasaownsky! I know this is still in a draft state but I left some feedback.

File: src/utils/getNormalizedStateData.js (hunk 4 above)
Comment (nasaownsky) [reply src/utils/getNormalizedStateData.js:75]: Oh, this and below question comes from previous branch, didn't notice that. Fixed.

File: src/utils/generateJailsChartData.js (hunk 6 above)
Comment (nasaownsky) [reply src/utils/generateJailsChartData.js:2]: Should I fix this for all files in this PR?

File: src/utils/generateJailsChartData.js (hunk 7 above)
Comment (nasaownsky) [reply src/utils/generateJailsChartData.js:30]: Done

File: src/utils/generateJailsChartData.js (hunk 8 above)
Comment (nasaownsky) [reply src/utils/generateJailsChartData.js:45]: Moved the logic of `periods` into another func. The rest is a little bit different in mentioned functions. Tests will be added closer to the end of PR.

File: src/utils/generateJailsKeyInsightsData.js (hunk 9 above)
Comment (nasaownsky) [reply src/utils/generateJailsKeyInsightsData.js:140]: Done

File: src/App.js (hunk 2 above)
Comment (nasaownsky) [reply src/App.js:110]: Yes

File: src/App.js (hunk 1 above)
Comment (nasaownsky) [reply src/App.js:118]: I'm planning to make this dynamic in this PR.

File: src/__tests__/App.test.js (hunk 3 above)
Comment (nasaownsky) [reply src/__tests__/App.test.js:35]: Done

File: src/utils/chartPeriods.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+
+const chartPeriods = (data, datasets) => {
```
Comment (jessex) [inline src/utils/chartPeriods.js:18]: Let's add a docstring to this one.

File: src/utils/generateJailsChartData.js (hunk 6 above)
Comment (jessex) [reply src/utils/generateJailsChartData.js:2]: Yes please! And in future PRs, too, but don't worry about files that predate your PR.

File: src/utils/generateJailsChartData.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+ * @param data - normalized data (see @returns of getNormalizedStateData)
+ * @param metric - metric name for which we generate chart data
+ * @param counties {string[]} - counties for which we generate chart data
+ * @param countyLabels {string[]} - humanized county names
+ * @param countyCoverage - percentage of counties in the state reporting data
+ * @returns {{
+ * datasets: {
+ *   metric: string
+ *   county: string
+ *   coveredCounty: string
```
Comment (jessex) [inline src/utils/generateJailsChartData.js:36]: Reminder: countyCoverage

**Review by jessex — COMMENTED** [review]
Looking good! Took another quick pass and made two minor notes. I'll hold off on another pass until this is out of draft state.

**Comment by nasaownsky** [comment]
Beside tests I think this PR is ready for review. Please have a look at it, @jessex 
I'll work on tests tomorrow.

**Review by jessex — COMMENTED** [review]
Looks good to go, @nasaownsky! Will approve once the test coverage is in a good state.

**Comment by nasaownsky** [comment]
All is done, @jessex

**Review by jessex — APPROVED** [review]
Test coverage looks great. Thanks, @nasaownsky! Have a great weekend.

**Pull request merged**